pkg/cri/server/container_checkpoint.go

```diff
@@ -0,0 +1,165 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
+
+	crmetadata "github.com/checkpoint-restore/checkpointctl/lib"
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/plugin"
+	"github.com/containerd/containerd/runtime/linux/runctypes"
+	"github.com/containerd/containerd/runtime/v2/runc/options"
+	runtime "k8s.io/cri-api/pkg/apis/runtime/v1"
+)
+
+func (c *criService) CheckpointContainer(
+	ctx context.Context,
+	r *runtime.CheckpointContainerRequest,
+) (*runtime.CheckpointContainerResponse, error) {
+	start := time.Now()
+
+	// Kubernetes has the possibility to request a file system local
+	// checkpoint archive. If the given location starts with a '/' or
+	// does not contain any slashes this assumes a local file.
+	// Only slashes in the middle assumes a destination in the local image store.
```

Comment:
It might be good to add an example:

- Local file: `foo`, `/foo/bar`
- Image: `localhost:5000/foo`, `quay.io/foo/bar`

Reply:
added